File: solutions/java/src/main/java/com/epi/MergeSortedLists.java
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.
// @author Ivan Sharov

package com.epi;

import java.util.Random;

public class MergeSortedLists {
  //@include
  public static Node<Integer> mergeTwoSortedLinkedLists(
      Node<Integer> F, Node<Integer> L) {
    Node<Integer> dummyHead = new Node<>(0, null);
    Node<Integer> current = dummyHead;
    Node<Integer> p1 = F;
    Node<Integer> p2 = L;

    while (p1 != null && p2 != null) {
      if (p1.data <= p2.data) {
        current.next = p1;
        p1 = p1.next;
      } else {
        current.next = p2;
        p2 = p2.next;
      }
      current = current.next;
    }

    if (p1 != null) {
      current.next = p1;
    }
    if (p2 != null) {
      current.next = p2;
    }
    return dummyHead.next;
  }
  // @exclude

  public static void main(String[] args) {
    Random rnd = new Random();
    for (int times = 0; times < 10000; ++times) {
      Node<Integer> F = null;
      Node<Integer> L = null;
      int n, m;
      if (args.length == 2) {
        n = Integer.parseInt(args[0]);
        m = Integer.parseInt(args[1]);
      } else if (args.length == 1) {
        n = Integer.parseInt(args[0]);
        m = Integer.parseInt(args[0]);
      } else {
        n = rnd.nextInt(100);
        m = rnd.nextInt(100);
      }
      for (int i = n; i > 0; --i) {
        Node<Integer> temp = new Node<>(i, null);
        temp.next = F;
        F = temp;
      }
      for (int j = m; j > 0; --j) {
        Node<Integer> temp = new Node<>(j, null);
        temp.next = L;
        L = temp;
      }

      Node<Integer> sortedHead = mergeTwoSortedLinkedLists(F, L);
      int count = 0;
      int pre = Integer.MIN_VALUE;
      while (sortedHead != null) {
        assert (pre <= sortedHead.data);
        pre = sortedHead.data;
        sortedHead = sortedHead.next;
        ++count;
      }
      // Make sure the merged list have the same number of nodes as F and L.
      assert (count == n + m);
    }
  }
}

File: solutions/cpp/Merge_sorted_lists.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <limits>
#include <memory>
#include <random>
#include <string>

#include "./Linked_list_prototype.h"
#include "./Merge_sorted_lists.h"

using std::cout;
using std::default_random_engine;
using std::endl;
using std::make_shared;
using std::numeric_limits;
using std::random_device;
using std::stoi;
using std::uniform_int_distribution;

void SimpleTest() {
  shared_ptr<ListNode<int>> L1 = nullptr;
  shared_ptr<ListNode<int>> L2 = nullptr;
  auto result = MergeTwoSortedLists(L1, L2);
  assert(result == nullptr);

  L1 = make_shared<ListNode<int>>(ListNode<int>{123, nullptr});
  result = MergeTwoSortedLists(L1, L2);
  assert(result->data == 123);

  L2 = L1;
  L1 = nullptr;
  result = MergeTwoSortedLists(L1, L2);
  assert(result->data == 123);

  L1 = make_shared<ListNode<int>>(ListNode<int>{-123, nullptr});
  L2 = make_shared<ListNode<int>>(ListNode<int>{123, nullptr});
  result = MergeTwoSortedLists(L1, L2);
  assert(result->data == -123 && result->next->data == 123 &&
         result->next->next == nullptr);
}

int main(int argc, char* argv[]) {
  SimpleTest();
  default_random_engine gen((random_device())());
  for (int times = 0; times < 10000; ++times) {
    shared_ptr<ListNode<int>> F = nullptr, L = nullptr;
    int n, m;
    if (argc == 3) {
      n = stoi(argv[1]), m = stoi(argv[2]);
    } else if (argc == 2) {
      n = stoi(argv[1]), m = stoi(argv[1]);
    } else {
      uniform_int_distribution<int> dis(0, 99);
      n = dis(gen), m = dis(gen);
    }
    for (int i = n; i > 0; --i) {
      shared_ptr<ListNode<int>> temp =
          make_shared<ListNode<int>>(ListNode<int>{i, nullptr});
      temp->next = F;
      F = temp;
    }
    for (int j = m; j > 0; --j) {
      shared_ptr<ListNode<int>> temp =
          make_shared<ListNode<int>>(ListNode<int>{j, nullptr});
      temp->next = L;
      L = temp;
    }

    shared_ptr<ListNode<int>> sorted_head = MergeTwoSortedLists(F, L);
    int count = 0;
    int pre = numeric_limits<int>::min();
    while (sorted_head) {
      assert(pre <= sorted_head->data);
      pre = sorted_head->data;
      sorted_head = sorted_head->next;
      ++count;
    }
    // Make sure the merged list have the same number of nodes as F and L.
    assert(count == n + m);
  }
  return 0;
}
